In [52]:
import pandas as pd
import seaborn as sns
import statsmodels.formula.api as smf
import statsmodels.api as sm2
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn import metrics 
from sklearn.model_selection  import train_test_split
import numpy as np

%matplotlib inline

In [53]:
df= pd.read_csv('auto.csv', header=0)
df.head()

make  listing  price  mpg  repairs  weight  length  foreign
0   AMC        1   4099   22        3    2930     186        0
1   AMC        2   4749   17        3    3350     173        0
2   AMC        3   3799   22        3    2640     168        0
3  Audi        4   9690   17        5    2830     189        1
4  Audi        5   6295   23        3    2070     174        1

In [54]:
df.corr(method='pearson')

listing     price       mpg   repairs    weight    length   foreign
listing  1.000000 -0.031020  0.296802  0.077346 -0.117224 -0.066199  0.213887
price   -0.031020  1.000000 -0.438462  0.021804  0.556072  0.436035  0.083517
mpg      0.296802 -0.438462  1.000000  0.232893 -0.808161 -0.768052  0.400338
repairs  0.077346  0.021804  0.232893  1.000000 -0.299339 -0.199705  0.695465
weight  -0.117224  0.556072 -0.808161 -0.299339  1.000000  0.906536 -0.601067
length  -0.066199  0.436035 -0.768052 -0.199705  0.906536  1.000000 -0.494155
foreign  0.213887  0.083517  0.400338  0.695465 -0.601067 -0.494155  1.000000

In [55]:
# model 1
lm1 = smf.ols(formula='mpg ~  weight', data=df).fit()
lm1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.653
Model:                            OLS   Adj. R-squared:                  0.639
Method:                 Least Squares   F-statistic:                     45.19
Date:                Sun, 21 Jun 2020   Prob (F-statistic):           5.93e-07
Time:                        19:30:22   Log-Likelihood:                -63.171
No. Observations:                  26   AIC:                             130.3
Df Residuals:                      24   BIC:                             132.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     38.0665      2.611     14.578      0.000      32.677      43.456
weight        -0.0055      0.001     -6.722      0.000      -0.007      -0.004
==============================================================================
Omnibus:                       10.972   Durbin-Watson:                   1.992
Prob(Omnibus):                  0.004   Jarque-Bera (JB):                9.976
Skew:                           1.109   Prob(JB):                      0.00682
Kurtosis:                       5.070   Cond. No.                     1.48e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.48e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [56]:
# model 2
lm_multi = smf.ols(formula='mpg ~ weight + length', data=df).fit()
lm_multi.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.660
Model:                            OLS   Adj. R-squared:                  0.631
Method:                 Least Squares   F-statistic:                     22.34
Date:                Sun, 21 Jun 2020   Prob (F-statistic):           4.07e-06
Time:                        19:30:22   Log-Likelihood:                -62.904
No. Observations:                  26   AIC:                             131.8
Df Residuals:                      23   BIC:                             135.6
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     44.1374      9.181      4.807      0.000      25.144      63.130
weight        -0.0043      0.002     -2.181      0.040      -0.008      -0.000
length        -0.0521      0.075     -0.690      0.497      -0.208       0.104
==============================================================================
Omnibus:                        9.065   Durbin-Watson:                   1.967
Prob(Omnibus):                  0.011   Jarque-Bera (JB):                7.468
Skew:                           0.972   Prob(JB):                       0.0239
Kurtosis:                       4.764   Cond. No.                     5.15e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.15e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The R-square=0.660 suggests model 2 fit the data reasonably well.

Model 2 has a higher R-squared (0.660) than model 1, which means that model 2 provides a better fit to the data than model that only includes weight.

Model 2 has adjusted R-square=0.631, which means not all variables are contributing.

The F statistic test the null hypothesis that all of the regression coefficients are equal to zero. For model 2, the p-val of F statistic is 4.07e-06 which is < 0.05. This means we can reject the null hypothesis, and model 2 has some validity in fitting the data.

Generally, a significance level alpha = 0.05 is used. If p val of t is < 0.05, we can reject the null hypothesis that the regression coefficient is equal to zero, and the variable affects the response. For model 2, the contribution of weight is significiant since its p-val of t = 0.040 < 0.05. The contribution of length is't significiant since its p-val of t > 0.05.